config/management/operational/src/validator_state.rs

```diff
@@ -2,67 +2,123 @@
     /// Check if the consensus key held in secure storage matches
     /// that registered on-chain for the validator.
     pub consensus_key_match: Option<bool>,
+
+    /// If the consensus_key_match is true, check if the key is unique
+    pub consensus_key_unique: Option<bool>,
+
+    /// Check if the fullnode network key held in secure storage matches
+    /// that registered on-chain.
+    pub fullnode_network_key_match: Option<bool>,
+
+    /// Check if a validator is in the latest validator set on-chain.
+    pub in_validator_set: Option<bool>,
+
+    /// Check if the validator network key held in secure storage matches
+    /// that registered on-chain.
+    pub validator_network_key_match: Option<bool>,
 }
 
 impl VerifyValidatorStateResult {
     pub fn is_state_consistent(&self) -> bool {
-        self.in_validator_set == Some(true) && self.consensus_key_match == Some(true)
+        self.in_validator_set == Some(true)
+            && self.consensus_key_match == Some(true)
+            && self.consensus_key_unique == Some(true)
+            && self.validator_network_key_match == Some(true)
+            && self.fullnode_network_key_match == Some(true)
     }
 }
 
 #[derive(Debug, StructOpt)]
 pub struct VerifyValidatorState {
+    #[structopt(flatten)]
+    config: ConfigPath,
     #[structopt(long, required_unless = "config")]
     json_server: Option<String>,
     #[structopt(flatten)]
-    validator_config: diem_management::validator_config::ValidatorConfig,
+    validator_backend: ValidatorBackend,
 }
 
 impl VerifyValidatorState {
     pub fn execute(self) -> Result<VerifyValidatorStateResult, Error> {
         // Load the config, storage backend and create a json rpc client.
         let config = self
-            .validator_config
-            .config()?
-            .override_json_server(&self.json_server);
+            .config
+            .load()?
+            .override_json_server(&self.json_server)
+            .override_validator_backend(&self.validator_backend.validator_backend)?;
         let storage = config.validator_backend();
         let client = JsonRpcClientWrapper::new(config.json_server);
         let owner_account = storage.account_address(OWNER_ACCOUNT)?;
 
-        // Verify if the validator is in the set
-        let in_validator_set = client
-            .validator_set(None)?
-            .iter()
-            .any(|vi| vi.account_address() == &owner_account);
+        let validator_infos = client.validator_set(None)?;
+
+        let mut result = VerifyValidatorStateResult {
+            consensus_key_match: None,
+            consensus_key_unique: None,
+            fullnode_network_key_match: None,
+            in_validator_set: None,
+            validator_network_key_match: None,
+        };
 
-        // TODO(khiemngo): consider return early if the validator is not in the set
+        // Verify if the validator is in the set.
+        // If it is not, return result and no need to do other checks.
```

Comment:
See my comment above, but I think we should try to do as many checks as possible (that make sense). This will help when debugging, e.g., think about a validator who has just joined. They may not be in the set yet, but they should still be able to verify that their keys match on-chain, etc, right? Why only wait until they are in the set before allowing them to verify local vs on-chain state

Reply:
Yeah it all makes sense now! I updated to proceed with other checks even if the node is not in the set.